octopus/src/test/scala/octopus/AsyncValidationRulesSpec.scala

```diff
@@ -0,0 +1,157 @@
+package octopus
+
+import java.io.IOException
+
+import octopus.dsl._
+import octopus.example.domain.{Email, User}
+import octopus.syntax._
+import org.scalatest.concurrent.{IntegrationPatience, ScalaFutures}
+import org.scalatest.{AsyncWordSpec, MustMatchers}
+
+import scala.concurrent.Future
+
+
+class AsyncValidationRulesSpec extends AsyncWordSpec
+  with ScalaFutures
+  with Fixtures
+  with IntegrationPatience
+  with MustMatchers {
+
+  private def userThrowNonFatal(user: User): Future[Boolean] = Future.failed(new Exception(Exception_handled_during_validation))
+  private def validateUserEither(user: User): Future[Either[String, Boolean]] = Future.successful { user.email match {
+    case Email(address) if address == email_Valid.address => Right(true)
+    case Email(address) if address == email_Valid_Long.address => Right(false)
+    case _ => Left(Email_validated_left_case)
+  }}
+  private def validateUserOption(user: User): Future[Option[Boolean]] = Future.successful { user.email match {
+    case Email(address) if address == email_Valid.address => Some(true)
+    case Email(address) if address == email_Valid_Long.address => Some(false)
+    case _ => None
+  }}
+
+  private val Email_does_not_contain_a = "Email does not contain a"
+  private val User_Invalid = "Invalid user"
+  private val Exception_handled_during_validation = "Exception handled during validation"
+  private val Email_invalid = "Invalid email"
+  private val Email_validated_left_case = "Invalid email left case"
+  private val User_validated_none_option = "Invalid user none option"
+
+  private val userValidator = Validator[User].async
+
+  "AsyncValidationRules" when {
+
+    "Simple email validator" should {
+
+      implicit val userUniqueEmailValidator = userValidator
+        .ruleField('email, (email: Email) => Future.successful(email.address.contains("a")), Email_does_not_contain_a)
+
+      "accept proper email" in {
+        user_Valid2.isValidAsync.map(_ mustBe true)
+      }
+
+      "reject invalid email" in {
+
+        val expectedValidationError = ValidationError(
+          message = Email_does_not_contain_a,
+          path = FieldPath(List(FieldLabel('email)))
+        )
+
+        user_Valid.isValidAsync.map(_ mustBe false)
+        user_Valid.validateAsync.map(_.errors must contain (expectedValidationError))
+      }
+    }
+
+    "Throwing email validator" should {
+
+      implicit val userUniqueThrowingValidator = userValidator
+        .ruleField('email, (_: Email) => Future.failed(new IOException()), Email_does_not_contain_a)
```

Comment:
let's use syntax with selectors (`_.email` instead of `'email`). Not sure how long we should support the latter.